File: sdk/ruby/lib/velocity_sdk/migrate.rb
# frozen_string_literal: true

# Velocity Ruby Migration Tool
#
# Scans a Ruby codebase for Temporal, Restate, or DBOS workflow patterns
# and converts them to Velocity Ruby SDK workflows.
#
# Usage:
#   ruby sdk/ruby/lib/velocity_sdk/migrate.rb --src ./my_project --from temporal
#   ruby sdk/ruby/lib/velocity_sdk/migrate.rb --src workflow.rb --from auto
#   ruby sdk/ruby/lib/velocity_sdk/migrate.rb --detect ./my_project

module VelocitySDK
  module Migrate
    # ─── Pattern Definition ────────────────────────────────────────────────

    MigrationPattern = Struct.new(:name, :source_pattern, :target_template, :source_framework)

    # ─── Temporal → Velocity Patterns ──────────────────────────────────────

    TEMPORAL_PATTERNS = [
      MigrationPattern.new('temporal-require-workflow',
        /require\s+['"]temporal\/workflow['"]/,
        "require 'velocity_sdk'"),
      MigrationPattern.new('temporal-require-activity',
        /require\s+['"]temporal\/activity['"]/,
        "require 'velocity_sdk'"),
      MigrationPattern.new('temporal-require-client',
        /require\s+['"]temporal\/client['"]/,
        "require 'velocity_sdk/client'"),
      MigrationPattern.new('temporal-require-worker',
        /require\s+['"]temporal\/worker['"]/,
        "require 'velocity_sdk/worker'"),

      # Module/class replacements
      MigrationPattern.new('temporal-workflow-module',
        /include\s+Temporal::Workflow/,
        'include VelocitySDK::Workflow'),
      MigrationPattern.new('temporal-activity-module',
        /include\s+Temporal::Activity/,
        'include VelocitySDK::Activity'),

      # Method call replacements
      MigrationPattern.new('temporal-execute-activity',
        /execute_activity\s*\(\s*['"]?(\w+)['"]?\s*/,
        'ctx.execute_activity(\1, '),
      MigrationPattern.new('temporal-sleep',
        /Temporal::Workflow\.sleep\s*\(/,
        'ctx.sleep('),
      MigrationPattern.new('temporal-side-effect',
        /Temporal::Workflow\.side_effect\s*\(/,
        'ctx.side_effect('),
      MigrationPattern.new('temporal-signal',
        /Temporal::Workflow::Signal/,
        'VelocitySDK::Signal'),
      MigrationPattern.new('temporal-query',
        /Temporal::Workflow::Query/,
        'VelocitySDK::Query'),

      # Client/Worker
      MigrationPattern.new('temporal-client-new',
        /Temporal::Client\.new\s*\(/,
        'VelocitySDK::VelocityClient.new('),
      MigrationPattern.new('temporal-worker-new',
        /Temporal::Worker\.new\s*\(/,
        'VelocitySDK::Worker.new('),
      MigrationPattern.new('temporal-register-workflow',
        /register_workflow\s*\(/,
        'register_workflow('),
      MigrationPattern.new('temporal-register-activity',
        /register_activity\s*\(/,
        'register_activity('),
      MigrationPattern.new('temporal-start-workflow',
        /client\.execute_workflow\s*\(/,
        'client.execute_workflow('),
    ].freeze

    # ─── Restate → Velocity Patterns ───────────────────────────────────────

    RESTATE_PATTERNS = [
      MigrationPattern.new('restate-require',
        /require\s+['"]restate['"]/,
        "require 'velocity_sdk'"),
      MigrationPattern.new('restate-module',
        /include\s+Restate::/,
        'include VelocitySDK::'),
      MigrationPattern.new('restate-ctx-run',
        /context\.run\s*\(/,
        'ctx.execute_activity('),
      MigrationPattern.new('restate-ctx-call',
        /context\.call\s*\(/,
        'ctx.execute_activity('),
      MigrationPattern.new('restate-ctx-sleep',
        /context\.sleep\s*\(/,
        'ctx.sleep('),
      MigrationPattern.new('restate-ctx-get',
        /context\.get\s*\(\s*['"](\w+)['"]/,
        "ctx.get_state('\\1')"),
      MigrationPattern.new('restate-ctx-set',
        /context\.set\s*\(\s*['"](\w+)['"]/,
        "ctx.set_state('\\1'"),
    ].freeze

    # ─── DBOS → Velocity Patterns ──────────────────────────────────────────

    DBOS_PATTERNS = [
      MigrationPattern.new('dbos-require',
        /require\s+['"]dbos['"]/,
        "require 'velocity_sdk'"),
      MigrationPattern.new('dbos-workflow-decorator',
        /extend\s+DBOS::Workflow/,
        'include VelocitySDK::Workflow'),
      MigrationPattern.new('dbos-transaction-decorator',
        /extend\s+DBOS::Transaction/,
        'include VelocitySDK::Activity'),
      MigrationPattern.new('dbos-sleep',
        /DBOS\.sleep\s*\(/,
        'ctx.sleep('),
      MigrationPattern.new('dbos-recv',
        /DBOS\.recv\s*\(/,
        'ctx.recv('),
      MigrationPattern.new('dbos-set-event',
        /DBOS\.set_event\s*\(/,
        'ctx.set_event('),
      MigrationPattern.new('dbos-get-event',
        /DBOS\.get_event\s*\(/,
        'ctx.get_event('),
    ].freeze

    ALL_PATTERNS = {
      'temporal' => TEMPORAL_PATTERNS,
      'restate'  => RESTATE_PATTERNS,
      'dbos'     => DBOS_PATTERNS,
    }.freeze

    # ─── Framework Detection ───────────────────────────────────────────────

    def self.detect_framework(content)
      scores = { 'temporal' => 0, 'restate' => 0, 'dbos' => 0 }
      evidence = { 'temporal' => [], 'restate' => [], 'dbos' => [] }

      # Temporal
      if content.match?(/temporal\/workflow/) || content.match?(/Temporal::Workflow/)
        scores['temporal'] += 3
        evidence['temporal'] << 'Temporal workflow import'
      end
      if content.match?(/Temporal::Activity/)
        scores['temporal'] += 2
        evidence['temporal'] << 'Temporal activity module'
      end
      if content.match?(/execute_activity/)
        scores['temporal'] += 1
        evidence['temporal'] << 'execute_activity call'
      end

      # Restate
      if content.match?(/require\s+['"]restate['"]/) || content.match?(/Restate::/)
        scores['restate'] += 3
        evidence['restate'] << 'Restate import'
      end
      if content.match?(/context\.run\s*\(/)
        scores['restate'] += 1
        evidence['restate'] << 'context.run()'
      end

      # DBOS
      if content.match?(/require\s+['"]dbos['"]/) || content.match?(/DBOS::/)
        scores['dbos'] += 3
        evidence['dbos'] << 'DBOS import'
      end
      if content.match?(/DBOS\.sleep/)
        scores['dbos'] += 1
        evidence['dbos'] << 'DBOS.sleep'
      end

      best = scores.max_by { |_, v| v }
      total = scores.values.sum
      confidence = total > 0 ? best[1].to_f / total : 0.0

      { framework: best[0], confidence: confidence, evidence: evidence[best[0]] }
    end

    # ─── File Migration ────────────────────────────────────────────────────

    def self.migrate_file(content, source_framework)
      result = { success: true, detected: '', transformations: 0, error: nil }

      if source_framework == 'auto'
        detection = detect_framework(content)
        result[:detected] = detection[:framework]
        if detection[:confidence] < 0.3
          return [content, { success: false, error: "Low confidence: #{detection[:framework]}" }]
        end
        source_framework = detection[:framework]
      else
        result[:detected] = source_framework
      end

      patterns = ALL_PATTERNS[source_framework]
      unless patterns
        return [content, { success: false, error: "Unknown framework: #{source_framework}" }]
      end

      migrated = content.dup
      count = 0
      patterns.each do |p|
        new_text = migrated.gsub(p.source_pattern, p.target_template)
        count += 1 if new_text != migrated
        migrated = new_text
      end
      result[:transformations] = count

      [migrated, result]
    end

    # ─── Project Scanner ───────────────────────────────────────────────────

    SKIP_DIRS = %w[vendor .git node_modules tmp log].freeze

    def self.scan_ruby_files(root_dir)
      files = []
      Dir.glob(File.join(root_dir, '**', '*.rb')).each do |f|
        next if SKIP_DIRS.any? { |d| f.include?("/#{d}/") }
        files << f
      end
      files
    end

    def self.has_workflow_content?(content)
      indicators = ['Temporal::', 'Restate::', 'DBOS::', 'temporal/', 'execute_activity', 'context.run']
      indicators.any? { |i| content.include?(i) }
    end

    # ─── Bulk Migration ────────────────────────────────────────────────────

    def self.bulk_migrate(source_dir, output_dir, from, dry_run: false)
      files = scan_ruby_files(source_dir)
      results = { total: files.size, migrated: 0, failed: 0, skipped: 0, details: [] }

      files.each do |file_path|
        content = File.read(file_path)
        unless has_workflow_content?(content)
          results[:skipped] += 1
          next
        end

        migrated, result = migrate_file(content, from)

        if result[:success] && !dry_run
          rel_path = file_path.sub("#{source_dir}/", '')
          out_path = File.join(output_dir, rel_path)
          FileUtils.mkdir_p(File.dirname(out_path))
          File.write(out_path, migrated)
          results[:migrated] += 1
        elsif result[:success]
          results[:migrated] += 1
        else
          results[:failed] += 1
        end

        results[:details] << result.merge(path: file_path)
      end

      results
    end
  end
end

# ─── CLI Entry Point ─────────────────────────────────────────────────────────

if __FILE__ == $PROGRAM_NAME
  require 'optparse'
  require 'fileutils'

  options = { from: 'auto' }
  OptionParser.new do |opts|
    opts.banner = "Velocity Ruby Migration Tool\n\nUsage: ruby migrate.rb [options]"
    opts.on('--src PATH', 'Source file or directory') { |v| options[:src] = v }
    opts.on('--from FRAMEWORK', 'Source: temporal, restate, dbos, auto') { |v| options[:from] = v }
    opts.on('--output PATH', '-o PATH', 'Output file or directory') { |v| options[:output] = v }
    opts.on('--dry-run', 'Detect without writing') { options[:dry_run] = true }
    opts.on('--detect', 'Detect framework in directory') { options[:detect] = true }
  end.parse!

  unless options[:src]
    $stderr.puts 'Error: --src is required'
    exit 1
  end

  if options[:detect]
    files = VelocitySDK::Migrate.scan_ruby_files(options[:src])
    puts "Scanning #{files.size} Ruby files in #{options[:src]}..."
    files.each do |f|
      content = File.read(f)
      d = VelocitySDK::Migrate.detect_framework(content)
      if d[:confidence] > 0.3
        rel = f.sub("#{options[:src]}/", '')
        puts "  #{rel}: #{d[:framework]} (#{(d[:confidence] * 100).round}%)"
      end
    end
    exit 0
  end

  if File.file?(options[:src])
    content = File.read(options[:src])
    migrated, result = VelocitySDK::Migrate.migrate_file(content, options[:from])
    unless result[:success]
      $stderr.puts "Failed: #{result[:error]}"
      exit 1
    end
    if options[:output]
      File.write(options[:output], migrated)
      puts "Written to: #{options[:output]}"
    else
      puts migrated
    end
    puts "\nDetected: #{result[:detected]}"
    puts "Transformations: #{result[:transformations]}"
    exit 0
  end

  if File.directory?(options[:src])
    output_dir = options[:output] || File.join(File.dirname(options[:src]), 'velocity-migrated')
    puts "Scanning: #{options[:src]}"
    puts "Output: #{options[:dry_run] ? '(dry run)' : output_dir}"
    puts "Source framework: #{options[:from]}\n\n"

    results = VelocitySDK::Migrate.bulk_migrate(
      options[:src], output_dir, options[:from], dry_run: options[:dry_run]
    )
    puts "Results:"
    puts "  Total: #{results[:total]}"
    puts "  Migrated: #{results[:migrated]}"
    puts "  Failed: #{results[:failed]}"
    puts "  Skipped: #{results[:skipped]}"
    results[:details].each do |r|
      status = r[:success] ? 'OK' : 'FAIL'
      puts "  [#{status}] #{r[:path]} (#{r[:detected]}, #{r[:transformations]} changes)"
    end
  end
end

File: sdk/rust/src/migrate.rs
//! Velocity Rust Migration Tool
//!
//! Scans a Rust codebase for Temporal, Restate, or DBOS workflow patterns
//! and converts them to Velocity Rust SDK workflows.
//!
//! Usage:
//!   cargo run --bin velocity-migrate -- --src ./my_project --from temporal
//!   cargo run --bin velocity-migrate -- --detect ./my_project

use std::fs;
use std::path::{Path, PathBuf};

/// A migration pattern with source regex and target replacement.
struct MigrationPattern {
    name: &'static str,
    source_pattern: &'static str,  // simple string match (regex crate optional)
    target_template: &'static str,
    source_framework: &'static str,
}

// ─── Temporal → Velocity Patterns ──────────────────────────────────────────

fn temporal_patterns() -> Vec<MigrationPattern> {
    vec![
        // Dependency/import replacements
        MigrationPattern {
            name: "temporal-dep-tokio",
            source_pattern: "temporal-client",
            target_template: "velocity-sdk",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-use-workflow",
            source_pattern: "use temporal_client::",
            target_template: "use velocity_sdk::",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-use-worker",
            source_pattern: "use temporal_worker::",
            target_template: "use velocity_sdk::worker::",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-use-activity",
            source_pattern: "use temporal_sdk::activity",
            target_template: "use velocity_sdk::activity",
            source_framework: "temporal",
        },

        // Function signature replacements
        MigrationPattern {
            name: "temporal-workflow-fn",
            source_pattern: "async fn workflow_fn",
            target_template: "async fn workflow_fn",
            source_framework: "temporal",
        },

        // Method call replacements
        MigrationPattern {
            name: "temporal-execute-activity",
            source_pattern: "ctx.execute_activity(",
            target_template: "ctx.execute_activity(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-sleep",
            source_pattern: "tokio::time::sleep(",
            target_template: "ctx.sleep(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-signal",
            source_pattern: "ctx.make_signal_channel(",
            target_template: "ctx.signal_channel(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-side-effect",
            source_pattern: "ctx.side_effect(",
            target_template: "ctx.side_effect(",
            source_framework: "temporal",
        },

        // Worker/Client
        MigrationPattern {
            name: "temporal-client-new",
            source_pattern: "Client::new(",
            target_template: "VelocityClient::new(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-worker-new",
            source_pattern: "Worker::new(",
            target_template: "VelocityWorker::new(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-register-workflow",
            source_pattern: "worker.register_workflow(",
            target_template: "worker.register_workflow(",
            source_framework: "temporal",
        },
        MigrationPattern {
            name: "temporal-register-activity",
            source_pattern: "worker.register_activity(",
            target_template: "worker.register_activity(",
            source_framework: "temporal",
        },
        // Search attributes
        MigrationPattern {
            name: "temporal-search-attributes",
            source_pattern: "workflow::search_attributes(",
            target_template: "ctx.search_attributes(",
            source_framework: "temporal",
        },
        // Memo
        MigrationPattern {
            name: "temporal-memo",
            source_pattern: "workflow::memo(",
            target_template: "ctx.memo(",
            source_framework: "temporal",
        },
        // Update handler
        MigrationPattern {
            name: "temporal-update-handler",
            source_pattern: "#[temporal::update]",
            target_template: "#[velocity::update]",
            source_framework: "temporal",
        },
        // Continue-as-new
        MigrationPattern {
            name: "temporal-continue-as-new",
            source_pattern: "workflow::continue_as_new(",
            target_template: "ctx.continue_as_new(",
            source_framework: "temporal",
        },
    ]
}

// ─── Restate → Velocity Patterns ───────────────────────────────────────────

fn restate_patterns() -> Vec<MigrationPattern> {
    vec![
        MigrationPattern {
            name: "restate-dep",
            source_pattern: "restate-sdk",
            target_template: "velocity-sdk",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-use",
            source_pattern: "use restate_sdk::",
            target_template: "use velocity_sdk::",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-ctx-run",
            source_pattern: "ctx.run(",
            target_template: "ctx.execute_activity(",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-ctx-sleep",
            source_pattern: "ctx.sleep(",
            target_template: "ctx.sleep(",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-ctx-get",
            source_pattern: "ctx.get::<",
            target_template: "ctx.get_state::<",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-ctx-set",
            source_pattern: "ctx.set(",
            target_template: "ctx.set_state(",
            source_framework: "restate",
        },
        MigrationPattern {
            name: "restate-service-handler",
            source_pattern: "#[restate::handler]",
            target_template: "#[velocity::workflow]",
            source_framework: "restate",
        },
        // Idempotency key
        MigrationPattern {
            name: "restate-idempotency-key",
            source_pattern: "ctx.idempotency_key()",
            target_template: "ctx.idempotency_key()",
            source_framework: "restate",
        },
        // Service client
        MigrationPattern {
            name: "restate-service-client",
            source_pattern: "restate::Client::new(",
            target_template: "velocity_sdk::Client::new(",
            source_framework: "restate",
        },
    ]
}

// ─── DBOS → Velocity Patterns ──────────────────────────────────────────────

fn dbos_patterns() -> Vec<MigrationPattern> {
    vec![
        MigrationPattern {
            name: "dbos-dep",
            source_pattern: "dbos-sdk",
            target_template: "velocity-sdk",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-use",
            source_pattern: "use dbos::",
            target_template: "use velocity_sdk::",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-workflow-attr",
            source_pattern: "#[dbos::workflow]",
            target_template: "#[velocity::workflow]",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-transaction-attr",
            source_pattern: "#[dbos::transaction]",
            target_template: "#[velocity::activity]",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-sleep",
            source_pattern: "dbos::sleep(",
            target_template: "ctx.sleep(",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-recv",
            source_pattern: "dbos::recv(",
            target_template: "ctx.recv(",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-set-event",
            source_pattern: "dbos::set_event(",
            target_template: "ctx.set_event(",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-get-event",
            source_pattern: "dbos::get_event(",
            target_template: "ctx.get_event(",
            source_framework: "dbos",
        },
        // Queue operations
        MigrationPattern {
            name: "dbos-queue-enqueue",
            source_pattern: "dbos::enqueue(",
            target_template: "ctx.enqueue(",
            source_framework: "dbos",
        },
        MigrationPattern {
            name: "dbos-queue-dequeue",
            source_pattern: "dbos::dequeue(",
            target_template: "ctx.dequeue(",
            source_framework: "dbos",
        },
        // HTTP handler
        MigrationPattern {
            name: "dbos-http-handler",
            source_pattern: "#[dbos::http_handler]",
            target_template: "#[velocity::http_handler]",
            source_framework: "dbos",
        },
    ]
}

// ─── Framework Detection ─────────────────────────────────────────────────────

pub struct DetectionResult {
    pub framework: String,
    pub confidence: f64,
    pub evidence: Vec<String>,
}

pub fn detect_framework(content: &str) -> DetectionResult {
    let mut scores: Vec<(&str, i32, Vec<String>)> = vec![
        ("temporal", 0, vec![]),
        ("restate", 0, vec![]),
        ("dbos", 0, vec![]),
    ];

    // Temporal checks
    if content.contains("temporal-client") || content.contains("temporal_client") {
        scores[0].1 += 3;
        scores[0].2.push("Temporal SDK dependency".into());
    }
    if content.contains("temporal_sdk") {
        scores[0].1 += 2;
        scores[0].2.push("temporal_sdk import".into());
    }
    if content.contains("ctx.execute_activity") {
        scores[0].1 += 1;
        scores[0].2.push("execute_activity call".into());
    }
    if content.contains("workflow::search_attributes") || content.contains("workflow::continue_as_new") {
        scores[0].1 += 1;
        scores[0].2.push("Temporal advanced workflow API".into());
    }
    if content.contains("#[temporal::update]") {
        scores[0].1 += 1;
        scores[0].2.push("Temporal update handler".into());
    }

    // Restate checks
    if content.contains("restate-sdk") || content.contains("restate_sdk") {
        scores[1].1 += 3;
        scores[1].2.push("Restate SDK dependency".into());
    }
    if content.contains("#[restate::") {
        scores[1].1 += 2;
        scores[1].2.push("Restate attribute".into());
    }
    if content.contains("ctx.run(") {
        scores[1].1 += 1;
        scores[1].2.push("ctx.run() call".into());
    }
    if content.contains("ctx.idempotency_key") || content.contains("restate::Client") {
        scores[1].1 += 1;
        scores[1].2.push("Restate idempotency/client".into());
    }

    // DBOS checks
    if content.contains("dbos-sdk") || content.contains("use dbos::") {
        scores[2].1 += 3;
        scores[2].2.push("DBOS SDK dependency".into());
    }
    if content.contains("#[dbos::") {
        scores[2].1 += 2;
        scores[2].2.push("DBOS attribute".into());
    }
    if content.contains("dbos::sleep") {
        scores[2].1 += 1;
        scores[2].2.push("dbos::sleep call".into());
    }
    if content.contains("dbos::enqueue") || content.contains("#[dbos::http_handler]") {
        scores[2].1 += 1;
        scores[2].2.push("DBOS queue/HTTP handler".into());
    }

    // Find best match
    let (best_fw, best_score, best_evidence) = scores
        .iter()
        .max_by_key(|(_, score, _)| *score)
        .map(|(fw, score, ev)| (*fw, *score, ev.clone()))
        .unwrap_or(("temporal", 0, vec![]));

    let total: i32 = scores.iter().map(|(_, s, _)| s).sum();
    let confidence = if total > 0 {
        best_score as f64 / total as f64
    } else {
        0.0
    };

    DetectionResult {
        framework: best_fw.to_string(),
        confidence,
        evidence: best_evidence,
    }
}

// ─── File Migration ──────────────────────────────────────────────────────────

pub struct FileResult {
    pub source_path: String,
    pub success: bool,
    pub error: Option<String>,
    pub detected_framework: String,
    pub transformations: usize,
}

pub fn migrate_file(content: &str, source_framework: &str) -> (String, FileResult) {
    let mut result = FileResult {
        source_path: String::new(),
        success: true,
        error: None,
        detected_framework: String::new(),
        transformations: 0,
    };

    let framework = if source_framework == "auto" {
        let detection = detect_framework(content);
        result.detected_framework = detection.framework.clone();
        if detection.confidence < 0.3 {
            result.success = false;
            result.error = Some(format!(
                "Low confidence: {} ({:.2})",
                detection.framework, detection.confidence
            ));
            return (content.to_string(), result);
        }
        detection.framework
    } else {
        result.detected_framework = source_framework.to_string();
        source_framework.to_string()
    };

    let patterns = match framework.as_str() {
        "temporal" => temporal_patterns(),
        "restate" => restate_patterns(),
        "dbos" => dbos_patterns(),
        _ => {
            result.success = false;
            result.error = Some(format!("Unknown framework: {}", framework));
            return (content.to_string(), result);
        }
    };

    let mut migrated = content.to_string();
    let mut count = 0;
    for p in &patterns {
        if migrated.contains(p.source_pattern) {
            migrated = migrated.replace(p.source_pattern, p.target_template);
            count += 1;
        }
    }
    result.transformations = count;

    (migrated, result)
}

// ─── Project Scanner ─────────────────────────────────────────────────────────

const SKIP_DIRS: &[&str] = &["target", ".git", "node_modules"];

pub fn scan_rust_files(root_dir: &Path) -> Vec<PathBuf> {
    let mut files = Vec::new();
    scan_dir(root_dir, &mut files);
    files
}

fn scan_dir(dir: &Path, files: &mut Vec<PathBuf>) {
    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };

    for entry in entries.flatten() {
        let path = entry.path();
        if path.is_dir() {
            let name = entry.file_name().to_string_lossy().to_string();
            if SKIP_DIRS.contains(&name.as_str()) {
                continue;
            }
            scan_dir(&path, files);
        } else if path.is_file() {
            if let Some(ext) = path.extension() {
                if ext == "rs" {
                    files.push(path);
                }
            }
        }
    }
}

pub fn has_workflow_content(content: &str) -> bool {
    let indicators = [
        "temporal", "restate", "dbos",
        "execute_activity", "ctx.run(",
        "#[restate::", "#[dbos::",
    ];
    indicators.iter().any(|i| content.contains(i))
}

// ─── CLI Entry Point ─────────────────────────────────────────────────────────

fn main() {
    let args: Vec<String> = std::env::args().collect();

    let mut src: Option<String> = None;
    let mut from = "auto".to_string();
    let mut output: Option<String> = None;
    let mut dry_run = false;
    let mut detect = false;

    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--src" => { i += 1; src = args.get(i).cloned(); }
            "--from" => { i += 1; from = args.get(i).cloned().unwrap_or("auto".into()); }
            "--output" | "-o" => { i += 1; output = args.get(i).cloned(); }
            "--dry-run" => dry_run = true,
            "--detect" => detect = true,
            "--help" | "-h" => {
                println!("Velocity Rust Migration Tool\n");
                println!("Usage:");
                println!("  --src <file|dir>     Source file or directory");
                println!("  --from <framework>   Source: temporal, restate, dbos, auto");
                println!("  --output <path>      Output file or directory");
                println!("  --dry-run            Detect without writing");
                println!("  --detect             Detect framework in directory");
                return;
            }
            _ => {}
        }
        i += 1;
    }

    let src = src.unwrap_or_else(|| {
        eprintln!("Error: --src is required");
        std::process::exit(1);
    });

    // Mode: detect
    if detect {
        let src_path = Path::new(&src);
        if !src_path.is_dir() {
            eprintln!("Error: --detect requires a directory");
            std::process::exit(1);
        }
        let files = scan_rust_files(src_path);
        println!("Scanning {} Rust files in {}...", files.len(), src);
        for f in &files {
            if let Ok(content) = fs::read_to_string(f) {
                let d = detect_framework(&content);
                if d.confidence > 0.3 {
                    let rel = f.strip_prefix(src_path).unwrap_or(f);
                    println!(
                        "  {}: {} ({:.0}%) [{}]",
                        rel.display(),
                        d.framework,
                        d.confidence * 100.0,
                        d.evidence.join(", ")
                    );
                }
            }
        }
        return;
    }

    let src_path = Path::new(&src);

    // Mode: single file
    if src_path.is_file() {
        let content = fs::read_to_string(src_path).unwrap_or_else(|e| {
            eprintln!("Error reading file: {}", e);
            std::process::exit(1);
        });

        let (migrated, result) = migrate_file(&content, &from);
        if !result.success {
            eprintln!("Migration failed: {}", result.error.unwrap_or_default());
            std::process::exit(1);
        }

        if let Some(out) = output {
            fs::write(&out, &migrated).unwrap();
            println!("Written to: {}", out);
        } else {
            print!("{}", migrated);
        }
        println!("\nDetected: {}", result.detected_framework);
        println!("Transformations: {}", result.transformations);
        return;
    }

    // Mode: directory
    let output_dir = output.unwrap_or_else(|| {
        format!("{}/../velocity-migrated", src)
    });

    println!("Scanning: {}", src);
    println!("Output: {}", if dry_run { "(dry run)" } else { &output_dir });
    println!("Source framework: {}\n", from);

    let files = scan_rust_files(src_path);
    let mut migrated_count = 0;
    let mut failed_count = 0;
    let mut skipped_count = 0;

    for f in &files {
        let content = match fs::read_to_string(f) {
            Ok(c) => c,
            Err(_) => { failed_count += 1; continue; }
        };

        if !has_workflow_content(&content) {
            skipped_count += 1;
            continue;
        }

        let (migrated, result) = migrate_file(&content, &from);
        let rel = f.strip_prefix(src_path).unwrap_or(f);

        if result.success && !dry_run {
            let out_path = Path::new(&output_dir).join(rel);
            if let Some(parent) = out_path.parent() {
                fs::create_dir_all(parent).ok();
            }
            fs::write(&out_path, &migrated).ok();
            migrated_count += 1;
        } else if result.success {
            migrated_count += 1;
        } else {
            failed_count += 1;
        }

        let status = if result.success { "OK" } else { "FAIL" };
        println!(
            "  [{}] {} ({}, {} changes)",
            status,
            rel.display(),
            result.detected_framework,
            result.transformations
        );
    }

    println!("\nResults: {} files, {} migrated, {} failed, {} skipped",
        files.len(), migrated_count, failed_count, skipped_count);
}
